### DISCUSSION TITLE
Attention and iris augmentations

### DISCUSSION THREAD

**Discussion by saeidn95** [vladmandic/human#460] [Q&A]
Here are a couple of question about the mesh network argumentation.

a. the attention augmentation which is disabled by default (https://github.com/vladmandic/human/blob/1114014bfd6c114baf2d31942382a2b2dc065c7d/src/face/facemesh.ts#L115), augments the 468 points from the mesh with 232 additional  points that are produced by the five additional set of outputs to a total of 700. These additional 232 points do not have z-coordinates and get their z values averaged from the set of corresponding points from the 468 mesh outputs.    

I might be wrong but, but my understanding of the attention is to replace points around the eyes and lips from the mesh output  with the more accurate points from the five attention output. If I am correct I would expect replacement and not addition (except of the 5+5 iris points). This seems to be how the default iriis attention network work  works.(https://github.com/vladmandic/human/blob/1114014bfd6c114baf2d31942382a2b2dc065c7d/src/face/facemesh.ts#L117)

With iris augmentation on the other hand, only adds 10 new points iris points and replaces  old mesh points with the new ones. Shouldn't we do the same thing with attention?

b. Where do i get the attention network in tflite/onnx format?

**Comment by vladmandic** [comment]
yes, we should replace points and not add them, but i never liked how mediapipe implemented augmentation to start with so i never went too far with it.

the issue is that augmentation network is not really single network, its basically pipe of multiple models connected using proprietary functions - and those functions are only implemented as unofficially software functions for tfjs-webgl backend, they don't exist in actual tensorflow.

for that reason, attention model cannot run in nodejs (since functions dont exist in tfjs-node) or tflite.

reality is, mediapipe never intended attention model to run outside of mediapipe backend.

**Reply by saeidn95** [reply]
Well explained! 

When I plot the both original 468  points from mesh model output and the corrected additional 222 from the five set of attention outputs, for the points around the eyes and lips, the corrected attention points are lot closer to their intended positions.  Unfortunately, these additional points  do not  z coordinates. I wonder if I can borrow the z coordinate from their corresponding original 468 points. I don't think using the averages would be very useful. 

The iris network that you use comes with 3 coordinates, which is good. 

Can you please point me to origin of the iris network in Human, and where I may get network in tflite/onnx format?

**Reply by vladmandic** [reply]
if you do map attention points back to full face mesh, it would be great - and i don't see a major problem reusing z-coordinates.
for iris, this is a good starting point: https://github.com/PINTO0309/PINTO_model_zoo/tree/main/049_iris_landmark

**Reply by vladmandic** [reply]
and this is a conversation on those functions in attention pipeline i've mentioned: https://github.com/PINTO0309/PINTO_model_zoo/issues/143

**Reply by saeidn95** [reply]
Great! Thank you,

a. Looking at your comment in https://github.com/PINTO0309/PINTO_model_zoo/issues/143
"fyi, i've just added facemesh attention to https://vladmandic.github.io/human,
works nice but it is ~%25 slower than facemesh + iris models combined"

I find this comment interesting. Whatever way we look at it, from this comment I conclude that it is best to avoid the  use of "facemesh+attention" and instead use "facemesh" and "iris". With iris we also get the a z component for the iris points, and things also run faster too! So, there is no point to worry about the "facemesh+attention" 

b. I also notice that you make  double correction to the z components of the iris network,
by not just by replacing old with new, but adding half of the old values the new values  ( newCoords[index][2] + rawCoords[originalIndices[j]][2]) / 2), (https://github.com/vladmandic/human/blob/1114014bfd6c114baf2d31942382a2b2dc065c7d/src/face/iris.ts#L49).  

Did you arrive at this double correction by experimentation?

**Reply by vladmandic** [reply]
that's just a trick to reduce jumps/flicker when processing real time video input such as webcam
